### PR TITLE
Revise CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by bdice**
I have revised the `CONTRIBUTING.md` file to address several pieces that are out of date. I also revised a good portion of the text and updated external references. Finally, I wrapped the lines at 100 characters to align with other Markdown files in the C++ docs. I would prefer to adopt a convention of one sentence per line if reviewers agree, but went with the 100 character wrapping for now to be consistent with other docs.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (bdice) [inline CONTRIBUTING.md:280]: I deleted the sections about using gpuCI locally and building Docker containers, because I think both are out of date. The Docker builds were removed in #10069. It is my understanding that the gpuCI scripts have not been an accurate representation of the CI workflow for some time. @rapidsai/ops-codeowners if you have any guidance or input on this, please feel free to weigh in!

File: CONTRIBUTING.md
Comment (jrhemstad) [inline CONTRIBUTING.md:69]: This is mainly an academic curiosity, but info like this goes stale very quickly. I wonder if there is a way to programmatically query it from a cmake file? @robertmaynard any ideas?

File: CONTRIBUTING.md
Comment (jrhemstad) [inline CONTRIBUTING.md:82]: I don't think we have sub-modules anymore?

File: CONTRIBUTING.md
Comment (jrhemstad) [inline CONTRIBUTING.md:92]: This makes it seem like `conda` is required. Many users frequently complain about this and we've gone to great lengths to make it possible to build without conda. 

We should have different sections for "Here's how to build with just cmake" and "here's how to build with a conda environment".

File: CONTRIBUTING.md
Comment (shwina) [reply CONTRIBUTING.md:69]: CMake gets picked up from the conda env, FWIW..

File: CONTRIBUTING.md
Comment (brandon-b-miller) [inline CONTRIBUTING.md:88]: I don't seem to remember needing this for a while. Is this actually still needed?

File: CONTRIBUTING.md
Comment (jrhemstad) [inline CONTRIBUTING.md:40]: There should be a step in here about writing/updating benchmarks as needed.

File: CONTRIBUTING.md
Comment (shwina) [inline CONTRIBUTING.md:307]: Also `mypy`?

File: CONTRIBUTING.md
Comment (brandon-b-miller) [inline CONTRIBUTING.md:210]: It would be useful if there was an example of how to run just one single test, for c++ and python. I rarely work in the c++ layer and always have to poke around slack to figure out how to run just one c++ test when needed.

**Comment by jrhemstad** [comment]
@bdice see also https://github.com/rapidsai/cudf/issues/9251

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:69]: I would prefer to point to an external source of this information (maybe the conda environment) and not specify versions here at all. (Unfortunately, our conda env doesn't necessarily specify our _minimal_ requirements - just a working configuration that is highly pinned.) I thought about this but I don't have any clever ideas for automating it.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md]: This agrees with a separate comment about having no submodules. Fixed.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md]: Fixed.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:306]: Alternatively, I considered eliminating this entire section about black/isort/flake8 in favor of just describing how to use pre-commit. That is guaranteed to be up-to-date since it's exactly how we run these tools in CI. What are your thoughts on that?

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:209]: Yes! I started this PR specifically to address concerns like:
- running a single test
- running a single benchmark
- running all benchmarks
- building docs

However, I think all of these topics would fit better in the C++ [Developer Guide](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/docs/DEVELOPER_GUIDE.md), [Benchmarking Guide](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/docs/BENCHMARKING.md), [Documentation Guide](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/docs/DOCUMENTATION.md), and [Testing Guide](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/docs/TESTING.md). Would it solve the problem if we cross-linked one or more of those documents here and did corresponding updates to those docs?

Comment (bdice) [reply CONTRIBUTING.md:209]: See also #9251 which has similar ideas for cross-linking.

File: CONTRIBUTING.md
Comment (robertmaynard) [reply CONTRIBUTING.md:69]: The root level CMakeLists.txt has a `cmake_minimum_required(VERSION 3.20.1 FATAL_ERROR)`. I expect we could run a script in our CI script process the makes sure the contributing.md and this line stay in sync.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:91]: Honestly, I have only built libcudf/cudf outside of rapids-compose once. I don't think I'm qualified to write/update these docs but I agree with the idea.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:41]: Fixed in 01d25ad6a0.

File: CONTRIBUTING.md
Comment (brandon-b-miller) [inline CONTRIBUTING.md:33]: it might help some newer contributors to clarify that this is supposed to happen in a branch in their own fork.

File: CONTRIBUTING.md
Comment (brandon-b-miller) [reply CONTRIBUTING.md:210]: As long as it makes it in somewhere I'm happy.

File: CONTRIBUTING.md
Comment (galipremsagar) [inline CONTRIBUTING.md:317]: Most of the times as a python developer, I run into gcc compilation issues. And when I do this is the place I would look at does my env have the right compatible gcc versions, since we are removing this what is the alternative place we can look at?

File: CONTRIBUTING.md
Comment (brandon-b-miller) [inline CONTRIBUTING.md:97]: I'd also consider noting that this file is updated fairly often. It's easy for it to fall out of sync since we upgrade things like arrow fairly often and can cause somewhat mystifying build errors if you don't already know it could be a problem.

File: CONTRIBUTING.md
Comment (shwina) [reply CONTRIBUTING.md:307]: Yeah, I think something along the lines of [this](https://docs.dask.org/en/latest/develop.html#code-formatting) would be great

Comment (bdice) [reply CONTRIBUTING.md:307]: Love it. I'll adopt similar wording here.

Comment (bdice) [reply CONTRIBUTING.md]: Done in 039fc0640e.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md]: That's a very good catch. This is commonly forgotten and results in PRs from the branch `username/branch-22.06`, which gets messy...

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:317]: This list was supposed to be Docker parameters but Docker builds were removed in #10069. This document lists compiler support above for `gcc 9.3+`. (However, that's not entirely true because we don't have compatibility with gcc 11 yet...)

We could try to expand the list of requirements here to mention Python versions, but this document has a tendency to become outdated very quickly. We only updated the Python package specifiers in #10579 (April 4, after [dropping Python 3.7 on February 3](https://docs.rapids.ai/notices/rsn0014/)), so I'm hesitant to add another source of truth here.

I would vote to take no action on this for now. I think this is better suited for a broader overhaul of both the C++/Python developer docs as we've proposed for some time now.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:99]: Do you like the change in 5c967782cf?

File: CONTRIBUTING.md
Comment (jrhemstad) [reply CONTRIBUTING.md:94]: I think this is worth putting the effort into as it likely the most valuable change we can make to this document. Building without conda is one of the most frequent user questions/requests we get. 

See 
https://github.com/rapidsai/cudf/issues/2770
https://github.com/rapidsai/cudf/issues/6979

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:72]: @jrhemstad Is an action item needed here to unblock/approve this PR or pursue in a follow-up PR? I'm inclined to punt on this as nice to have but out of scope.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:94]: @jrhemstad Could we defer this to a later PR? I do think this would be a significant improvement (and may illuminate some challenges with installing from source as mentioned in those issues), but I think it's out of scope for this PR.

File: CONTRIBUTING.md
Comment (bdice) [reply CONTRIBUTING.md:216]: I am going to leave this conversation open since it contains action items for future PRs. However, it would be nice to get a PR approval if you think the rest is good to go.

**Review by brandon-b-miller — APPROVED** [review]

**Comment by bdice** [comment]
@jrhemstad @brandon-b-miller I tagged you both for review since you provided the most comments in prior iterations - I'd like to merge this as-is and defer follow-up work to later PRs. This PR gets the formatting to be consistent, which will make the diffs of future changes easier to handle. Thanks!

File: CONTRIBUTING.md
Comment (jrhemstad) [reply CONTRIBUTING.md:94]: Minimum I'd like to see for this PR:
- Try building w/o conda and see what happens. If it works, great! Document that this is possible
- If it doesn't:
   -   Add some words to this document that states how a conda environment is currently required to satisfy build dependencies (these could be separated based on building libcudf vs cython/cudf python)
   - List what those dependencies are
   - File an issue to capture what work is needed to avoid requiring a conda environment for building libcudf

Comment (galipremsagar) [reply CONTRIBUTING.md:94]: > * Try building w/o conda and see what happens. If it works, great! Document that this is possible

Building w/o conda is totally possible. We previously had the steps inside a docker file to make this happen for the entire `cudf`, but was removed : https://github.com/rapidsai/cudf/pull/10069/files#diff-64993c0abc780bf9aca0d674016ebd25262ff0c83cb55824cc540455d60b2763

`docker_build/Dockerfile` is the file. Is this close to what you are looking for Jake? Some of the commands _could_ be outdated or unnecessary now but those should be helpful if you are looking to move those steps to here.

Comment (bdice) [reply CONTRIBUTING.md:94]: I deactivated all conda environments (`which python -> /usr/bin/python`) and attempted to run `build.sh`. It failed immediately:

```
-- NO RAPIDS_VERSION for CUDAToolkit
CMake Error at /usr/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.23/Modules/FindZLIB.cmake:120 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  build/_deps/rapids-cmake-src/rapids-cmake/find/package.cmake:89 (find_package)
  CMakeLists.txt:125 (rapids_find_package)
```

I estimate it will take me at least 3 hours to figure out how to acquire all the dependencies and build this without conda, and I will probably need to consult with others to make progress here. If you think it's a blocker for this PR, I can work on that further -- I just want to make it clear that it will be a significant time investment and that I feel the PR is otherwise ready. I aimed to scope this PR for eliminating outdated content, standardizing formatting, and making minor updates where appropriate, and this feels like it will require a substantial addition of new content.

Comment (bdice) [reply CONTRIBUTING.md:94]: Could we write the following?

> The build dependencies of libcudf and the cudf Python library are easiest to satisfy through the use of a conda environment. The list of dependencies can be found in `conda/environments/cudf_dev_cuda11.5.yml`. Building from source outside of a conda environment is not currently tested.

Comment (bdice) [reply CONTRIBUTING.md:94]: @galipremsagar Ah! I didn't see your comment while writing my responses above. That's super helpful. The scope/difficulty of documenting this is way lower if I can start from that Dockerfile and update it. In that case, I am much happier about trying to do this in the current PR. 💯

Comment (galipremsagar) [reply CONTRIBUTING.md:94]: > @galipremsagar Ah! I didn't see your comment while writing my responses above. That's super helpful. The scope/difficulty of documenting this is way lower if I can start from that Dockerfile and update it. In that case, I am much happier about trying to do this in the current PR. 💯

Yeah, most of the heavy lifting & testing was done up to a certain previous version of `cudf`. But happy to help if you run into issues using that script now.

Comment (jrhemstad) [reply CONTRIBUTING.md:94]: https://github.com/rapidsai/cudf/pull/10644#discussion_r851330772

This is fine too.

Comment (robertmaynard) [reply CONTRIBUTING.md:94]: Having the following packages would allow you to build cudf C++ library from scratch presuming you have an internet connection, as the rest of the dependencies will be downloaded as needed.

- cuda toolkit 
- cmake ( from https://apt.kitware.com/ )
- build-essential
- g++-9
- libpython3-dev
- libssl-dev 
- libz-dev

Comment (galipremsagar) [reply CONTRIBUTING.md:94]: I think it would be a good idea to maintain a non-conda build instructions, because right after we removed the docker file merlin team reached out with this request: https://github.com/rapidsai/cudf/issues/10301#issuecomment-1041868347

Comment (robertmaynard) [reply CONTRIBUTING.md:94]: Here is a minimal build for c++ ( libcudf )

https://gist.github.com/robertmaynard/93839d326f3b179cd04cbe1b371e229f

Comment (bdice) [reply CONTRIBUTING.md:98]: @jrhemstad I was able to build the minimal Dockerfile that @robertmaynard linked. I committed the list of minimal dependencies in c34658259194e26e6d46563536b55781c5a007eb and this PR is ready for your review. I was able to build without `software-properties-common` so I did not include that one.

Do we need to include a special note about `--cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"`? (Or disable that option by default, like the other disabled-by-default I/O features?)

Comment (bdice) [reply CONTRIBUTING.md:98]: Also, we list `gcc 9.3+` in the "General requirements." Do we need to list `build-essential` in the core dependencies as well, or is it reasonable to think that listing a compiler requirement will necessitate the other components of `build-essential`?

Comment (robertmaynard) [reply CONTRIBUTING.md:98]: Our base docker file brings in `build-essential`, but we can list it here so that users know it is required when building from a clean machine ( non-docker ).

Comment (robertmaynard) [reply CONTRIBUTING.md:98]: > Do we need to include a special note about --cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"

Good question. I am noticing some difference in cmake search paths in the dockerfile compared to base Ubuntu. We might be able to enable s3 once I figure out what is going wrong.

File: CONTRIBUTING.md
Comment (jrhemstad) [reply CONTRIBUTING.md:76]: It was just a "nice to have" curiosity. Definitely not a blocker.

File: CONTRIBUTING.md
Comment (jrhemstad) [reply CONTRIBUTING.md:98]: @jlowe would you mind reviewing the section about building w/o conda and verifying that this documentation would be sufficient to reproduce what you've been doing for jar builds?

Comment (bdice) [reply CONTRIBUTING.md:98]: Oops, I put this information in an awkward place. At first I couldn't decide how to organize this information and then I forgot to change it once I made a decision. I am going to put these dependencies between "Create the build environment" and "Build cuDF from source." The instructions will be the same, calling `build.sh`, so we just need to list the dependencies above that.

Comment (robertmaynard) [reply CONTRIBUTING.md:98]: > Do we need to include a special note about `--cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"`? (Or disable that option by default, like the other disabled-by-default I/O features?)

So we have a couple of options:

- We can enable arrow by also installing libcurl4-openssl-dev and updating the  `get_arrow.cmake` script to pass `-DAWSSDK_SOURCE=AUTO`
- Or we can document why we are disabling S3 ( less requirements? )

I am hestiant to change the default value of `CUDF_ENABLE_ARROW_S3` as that will be a breaking change to anyone building cudf presuming enabled by default ( users of `build.sh` or people directly calling `cmake` ).

As for why we need to pass `AWSSDK_SOURCE=AUTO` that is required for Arrow's build to work correctly when you don't have arrow already installled. This tells it you are okay with it building the dependency from source when it can't find it locally.

Comment (bdice) [reply CONTRIBUTING.md:103]: Fixed in f674f7f8a1. @jlowe Does this make more sense now?

Comment (jlowe) [reply CONTRIBUTING.md:103]: Yep, this makes more sense. Thanks!

**Review by jlowe — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**
